### PR TITLE
Moving sharp ib v2

### PR DESCRIPTION
**Pull request by Luckabarbeau**
Add the capacity to simulate moving IB particles using an implicit Euler shame. Add modification to the linear and non-linear solver to make the process more robust.

### CODE REVIEW

File: applications_tests/gls_sharp_navier_stokes_2d/moving_ib_gls.output
Comment (blaisb) [inline applications_tests/gls_sharp_navier_stokes_2d/moving_ib_gls.output:21]: p here is position?

File: applications_tests/gls_sharp_navier_stokes_2d/moving_ib_gls.prm
Comment (blaisb) [inline applications_tests/gls_sharp_navier_stokes_2d/moving_ib_gls.prm:46]:
```suggestion
```

File: applications_tests/gls_sharp_navier_stokes_2d/moving_ib_gls.prm
Comment (blaisb) [inline applications_tests/gls_sharp_navier_stokes_2d/moving_ib_gls.prm:80]:
```suggestion
```

File: include/core/ib_particle.h
Comment (blaisb) [inline include/core/ib_particle.h:45]: This comment needs to make more sense than that...

File: include/core/ib_particle.h
Comment (blaisb) [inline include/core/ib_particle.h:31]: Add more documentations for the parameters. What is the meaning of last force, forces, last position, etc. You should add a paragraph that describes the content and the role of the class.

File: include/solvers/gls_sharp_navier_stokes.h
Comment (blaisb) [inline include/solvers/gls_sharp_navier_stokes.h:83]:
```suggestion
  integrate_particles();
```

File: include/solvers/gls_sharp_navier_stokes.h
Comment (blaisb) [inline include/solvers/gls_sharp_navier_stokes.h:79]: Add header comment to the function that explains what it does.

File: include/solvers/gls_sharp_navier_stokes.h
Comment (blaisb) [inline include/solvers/gls_sharp_navier_stokes.h:82]: Add header to function that explains what it does.

File: include/solvers/gls_sharp_navier_stokes.h
Comment (blaisb) [inline include/solvers/gls_sharp_navier_stokes.h:84]: Add header to the function that explain what it does.

File: source/core/parameters.cc
Comment (blaisb) [inline source/core/parameters.cc:962]: Why uniform rotation inertia? Should it not be moment of inertia? Or you want to say uniform because here it is not a tensor?

File: source/core/parameters.cc
Comment (blaisb) [inline source/core/parameters.cc:1128]:
```suggestion
              particles[i].mass               = 4.0/3.0*3.14159265359*particles[i].radius*particles[i].radius*particles[i].radius*prm.get_double("density");
```

Comment (blaisb) [inline source/core/parameters.cc:1128]: You should use the PI from deal.II.
https://www.dealii.org/current/doxygen/deal.II/namespacenumbers.html

File: source/solvers/gls_navier_stokes.cc
Comment (blaisb) [inline source/solvers/gls_navier_stokes.cc:1376]: Add more comments here that explain what this does. Otherwise it's a bit confusing for a new person.

File: include/core/parameters.h
Comment (blaisb) [inline include/core/parameters.h:567]: Why is gravity a double? What happens if gravity is specified in diagonal?

File: source/solvers/gls_sharp_navier_stokes.cc
Comment (blaisb) [inline source/solvers/gls_sharp_navier_stokes.cc:110]: Why not create a default constructor for your particle class instead that sets those value? It would be a lot better to have a default constructor than to rely on the solver itself to set the particles. Encapsulation young man!

File: source/solvers/gls_sharp_navier_stokes.cc
Comment (blaisb) [inline source/solvers/gls_sharp_navier_stokes.cc:3524]: Spelling

File: source/solvers/gls_sharp_navier_stokes.cc
Comment (blaisb) [inline source/solvers/gls_sharp_navier_stokes.cc:3543]: spelling

File: source/solvers/gls_sharp_navier_stokes.cc
Comment (blaisb) [inline source/solvers/gls_sharp_navier_stokes.cc:3545]: Is this barrier necessary?

**Review by blaisb — COMMENTED** [review]
Some first wave of comments to apply. I'll need to recheck another time after that :)!.

File: applications_tests/gls_sharp_navier_stokes_2d/moving_ib_gls.output
Comment (Luckabarbeau) [reply applications_tests/gls_sharp_navier_stokes_2d/moving_ib_gls.output:21]: yeah

File: source/core/parameters.cc
Comment (Luckabarbeau) [reply source/core/parameters.cc:962]: Even though it's defined as a tensor in the solver, since the parameter only take a single value the moment of inertia is uniform.

File: source/solvers/gls_sharp_navier_stokes.cc
Comment (Luckabarbeau) [reply source/solvers/gls_sharp_navier_stokes.cc:3544]: dont think it's anymore i removed it

**Comment by blaisb** [comment]
@Luckabarbeau You need to resolve the comments above when they have been resolved. Otherwise it is difficult to keep track of what has or has not been done.

File: include/core/ib_particle.h
Comment (blaisb) [inline include/core/ib_particle.h:37]: Fix indentation
There are some spelling mistakes. Check spelling.
```suggestion
    // This class defines values related to a particle used for in the sharp interface immersed boundary.
    // Each particle defined will have these value used in the solver
    void initialise_all();
    void initialise_last();
```

Last line of comment is not clear.

File: include/core/ib_particle.h
Comment (blaisb) [inline include/core/ib_particle.h]: Why so many empty lines? To fix.

File: include/core/parameters.h
Comment (blaisb) [inline include/core/parameters.h]: Why?

File: include/solvers/gls_sharp_navier_stokes.h
Comment (blaisb) [inline include/solvers/gls_sharp_navier_stokes.h:71]: Fix the comments here. There is still this BB- TODO
Additionnaly, use present tense.

File: include/solvers/gls_sharp_navier_stokes.h
Comment (blaisb) [inline include/solvers/gls_sharp_navier_stokes.h:75]:
```suggestion
  // Evaluate the forces applied on each of the IB particles.
```

File: include/solvers/gls_sharp_navier_stokes.h
Comment (blaisb) [inline include/solvers/gls_sharp_navier_stokes.h:79]:
```suggestion
  // Modify the system matrix to impose IB condition
```

File: include/solvers/gls_sharp_navier_stokes.h
Comment (blaisb) [inline include/solvers/gls_sharp_navier_stokes.h:83]: Be more careful about spelling. This is important...
```suggestion
  // Write in  the ouput file the force , velocity , position of each of the particles
```

File: include/solvers/gls_sharp_navier_stokes.h
Comment (blaisb) [inline include/solvers/gls_sharp_navier_stokes.h:91]:
```suggestion
  // Store the solution for the particles dynamics parameters for integration. Defines the table to store the history of each of the particles
```

File: source/core/ib_particle.cc
Comment (blaisb) [inline source/core/ib_particle.cc:7]: Why an initialize function and not a constructor?

**Review by blaisb — COMMENTED** [review]
Still some comments to address.
Please resolve comments when they are addressed

File: source/core/ib_particle.cc
Comment (Luckabarbeau) [reply source/core/ib_particle.cc:7]: I have multiple initialize function for different parameters. But maybe using 1 constructor and 1 initialize could do the trick also.

File: include/core/ib_particle.h
Comment (blaisb) [inline include/core/ib_particle.h:32]: FIx spelling. "defines", etc. Make explanation clearer.

**Review by blaisb — COMMENTED** [review]
Some minor spelling/ comments to fix, but otherwise seems good to go. I think we will wait until #137 is merged before merging this one. Otherwise #137 will never finish :). Once #137 is merged you can ask @oguevremont  to help you rebase.

**Review by blaisb — APPROVED** [review]
Looking good. Ready for merge.

**Pull request merged**
